linera-core/src/client/mod.rs

```diff
@@ -1450,22 +1438,27 @@ where
         #[cfg(with_metrics)]
         let _latency = metrics::SYNCHRONIZE_CHAIN_STATE_LATENCY.measure_latency();
 
-        let mut futures = vec![];
-
-        for remote_node in validators {
-            let client = self.clone();
-            futures.push(async move {
-                client
-                    .try_synchronize_chain_state_from(remote_node, chain_id)
-                    .await
-            });
-        }
-
-        for result in future::join_all(futures).await {
-            if let Err(e) = result {
-                error!(?e, "Error synchronizing chain state");
-            }
-        }
+        let committee = self.local_committee().await?;
+        communicate_with_quorum(
+            validators,
+            &committee,
+            |_: &()| (),
+            |remote_node| {
+                let client = self.clone();
+                async move {
+                    client
+                        .try_synchronize_chain_state_from(&remote_node, chain_id)
```

Comment:
Just for the sake of my understanding – we're leaving "1 - quorum" number of tasks running in the background?

Reply:
Technically we're canceling those tasks.

But in the places where that wouldn't be safe (when handling certificates), we're now using `spawn` to make sure that these are still fully handled, including notifications, if they were started.

Reply:
Where does the cancelation happen? Inside `communicate_with_quorum` we create a future containing `execute` call (here `client.try_synchronize_chain_state_from`) for every member of the `validators`.

Reply:
Yes, and we turn the futures into a `FuturesUnordered` stream. But then we break out of the loop based on the quorum and/or timeout, effectively cancelling the futures that haven't completed yet.

Reply:
How are we canceling them? By the virtue of [`Drop::drop`-ing](https://docs.rs/futures-util/0.3.31/src/futures_util/stream/futures_unordered/mod.rs.html#575) the `FuturesUnordered`?

Reply:
Yes, exactly. They get dropped, but we probably already started polling them, so they are in the middle of execution.